sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -94,399 +47,145 @@
-      @Override
-      public PipelineOptions getPipelineOptions() {
-        return pipelineOptions;
-      }
-    };
+    this.startBundleContext =
+        this.context.doFn.new StartBundleContext() {
+          @Override
+          public PipelineOptions getPipelineOptions() {
+            return context.pipelineOptions;
+          }
+        };
     this.processContext = new ProcessBundleContext();
-    finishBundleContext = doFn.new FinishBundleContext() {
-      @Override
-      public PipelineOptions getPipelineOptions() {
-        return pipelineOptions;
-      }
-
-      @Override
-      public void output(OutputT output, Instant timestamp, BoundedWindow window) {
-        outputTo(mainOutputConsumers,
-            WindowedValue.of(output, timestamp, window, PaneInfo.NO_FIRING));
-      }
-
-      @Override
-      public <T> void output(TupleTag<T> tag, T output, Instant timestamp, BoundedWindow window) {
-        Collection<FnDataReceiver<WindowedValue<T>>> consumers = (Collection) outputMap.get(tag);
-        if (consumers == null) {
-          throw new IllegalArgumentException(String.format("Unknown output tag %s", tag));
-        }
-        outputTo(consumers,
-            WindowedValue.of(output, timestamp, window, PaneInfo.NO_FIRING));
-      }
-    };
+    finishBundleContext =
+        this.context.doFn.new FinishBundleContext() {
+          @Override
+          public PipelineOptions getPipelineOptions() {
+            return context.pipelineOptions;
+          }
+
+          @Override
+          public void output(OutputT output, Instant timestamp, BoundedWindow window) {
+            outputTo(
+                mainOutputConsumers,
+                WindowedValue.of(output, timestamp, window, PaneInfo.NO_FIRING));
+          }
+
+          @Override
+          public <T> void output(
+              TupleTag<T> tag, T output, Instant timestamp, BoundedWindow window) {
+            Collection<FnDataReceiver<WindowedValue<T>>> consumers =
+                (Collection) context.tagToConsumer.get(tag);
+            if (consumers == null) {
+              throw new IllegalArgumentException(String.format("Unknown output tag %s", tag));
+            }
+            outputTo(consumers, WindowedValue.of(output, timestamp, window, PaneInfo.NO_FIRING));
+          }
+        };
   }
 
   @Override
   public void startBundle() {
+    this.stateAccessor =
+        new FnApiStateAccessor(
+            context.pipelineOptions,
+            context.ptransformId,
+            context.processBundleInstructionId,
+            context.tagToSideInputSpecMap,
+            context.beamFnStateClient,
+            context.keyCoder,
+            (Coder<BoundedWindow>) context.windowCoder);
+
     doFnInvoker.invokeStartBundle(startBundleContext);
   }
 
   @Override
   public void processElement(WindowedValue<InputT> elem) {
     currentElement = elem;
+    stateAccessor.setCurrentElement(elem);
```

Comment:
What do you think about having a `current element` context object which we can share with the `DoFnRunner` and the `stateAccessor`? This way we only have one place to set the current element/window and to clear it.